### ISSUE TITLE
editing the list of fields to sanitize

### ISSUE THREAD

**Issue by wyattisimo** [getsentry/raven-ruby#295]
I'm perceiving a couple of issues with how the list of fields to sanitize is handled.

1. **The language is not clear.** When I say `config.sanitize_fields = %w(passwd poop monkeybutt)`, I intuit that those are the only fields that will be sanitized. In reality, those are added to an existing list of default fields.

2. **There's no option to change/replace the default list.** The default list is helpful, but I may have a legitimate need to exclude one or more of the default fields.

Ideas:
```ruby
# replacing the entire list
config.sanitize_fields = %w(passwd poop monkeybutt)

# adding fields to the list
config.sanitize_fields |= %w(passwd poop monkeybutt)
# or
config.sanitize_additional_fields = %w(passwd poop monkeybutt)
```

Thoughts?

**Comment by nateberkopec** [comment]
@dcramer Where can we find a list of what's sanitized *no matter what* on the Sentry server end?

**Comment by dcramer** [comment]
There's actually a setting to disable it, but theres no real public list. 

https://github.com/getsentry/sentry/blob/master/src/sentry/utils/data_scrubber.py#L38

**Comment by nateberkopec** [comment]
@wyattisimo can you take a look at #297?

**Comment by nateberkopec** [comment]
Based on how I feel after implementing #297, I'm feeling more strongly about changing the name of the config option to "sanitize_additional_fields" and keeping the stock behavior.

If you really think you have a good reason to override the default field sanitizations, you can probably monkeypatch the constant after it's loaded but before you start up the client.

**Comment by wyattisimo** [comment]
@nateberkopec I like `sanitize_additional_fields`.

You're right, there's not a strong use case for overriding the defaults if they stay basic and sensible. I think I just went there because the property was called `sanitize_fields` and it seemed that overriding was intuitively what it *should* do.

**Comment by wyattisimo** [comment]
@nateberkopec @dcramer Looking at this thread, and at #297, and thinking thoughts in my brain...  maybe the whole sanitize fields paradigm needs to be reassessed.

## La Ballad de Fields Desinfectado de el Fuego

What we currently have is a redundancy in field sanitizing logic between server and client. This results in a potential bug entry point on both server and client for what is essentially a single feature; it's a bit unwieldy, and may produce unexpected behavior.

If I choose not to sanitize data by not using the `SanitizeData` processor, that won't cut it. I also have to go to my project settings and uncheck the "Data Scrubber" box.

If I'm really concerned about sanitizing data, then I want to know what all the defaults are so I know what, if any, fields I need to add. But this means I have to look at the code for both the client and the server. Also, I have to *know* that I need to look in both places—honestly, before I saw @dcramer's comment above, I had no idea the server was sanitizing data, too.

Realistically, developers truly concerned with sanitizing will probably just end up sending a complete list of all the fields they want to sanitize to avoid any problems in case any of the defaults change in future updates. That's what I did when I first start using Sentry.

In my opinion, sanitizing should be handled by the client, not the server, for two main reasons:

1. Sanitizing on the client side means no sensitive data goes through the tubes.
2. The client knows the idiosyncrasies of the data—the server does not, and should therefore try to remain agnostic.

I don't think it's reasonable to kill the data scrubber on the server side at this point since lots of projects are relying on it to hide sensitive data. And since data is sent to Sentry over https, my reason (1) above is kind of meh.

So here's my suggestion: Expose Sentry's server-side data scrubbing via the API and allow the client to configure it. We could replace the `SanitizeData` processor in this gem with simply sending a list of sanitize fields to the Sentry API.

eh?

**Comment by nateberkopec** [comment]
@wyattisimo Thanks for writing all of your thoughts down. I'll let @dcramer parse and respond to all of your server-side concerns. I'm just the client guy here.

For the client, I think we could improve:

* [ ] The raven-ruby client documentation on the Wiki about this.
* [ ] An inline comment in the SanitizeData processor that either points you to the wiki or some other explanation of the sanitization of data that occurs on the Sentry server.

**Comment by dcramer** [comment]
So it's really complex for the server to communicate any kind of configuration to the client, especially because client's can be short lived etc.

If we were to ever do this, we'd do it through a proxy (we'll eventually release a binary that you can install on local systems to buffer requests).

Also we will never remove the server data scrubbers (as a default) as its more or less a "I understand the consequences, please allow me to accidentally store credit cards".

That said, I do think we might want to revisit the way we build sanitizers on clients. The class-based stuff is a bit cumbersome and gross, and arguably it'd be better if the user could just specify a list of regexp (or similar) and we apply them cleanly over the data.

**Comment by wyattisimo** [comment]
@dcramer Maybe I worded it in a way that sounded more complicated than it really is. Here's what I'm suggesting: https://github.com/getsentry/sentry/pull/1416

**Comment by wyattisimo** [comment]
@nateberkopec once that PR is released into the wild, you can just hit the project details endpoint with options: `sentry:scrub_data` and `sentry:sensitive_fields`

![BOOM](http://missashleypants.com/wp-content/uploads/2014/06/boom.gif)

**Comment by nateberkopec** [comment]
Ha! Awesome!

**Comment by nateberkopec** [comment]
Fixed upstream at getsentry/sentry#1416
